File: MomentumiOS/MomentumiOS/Presentation/Payments/PaymentSummaryContentViewModel.swift
//
//  PaymentSummaryContentViewModel.swift
//  MomentumiOS
//
//  Created by Mwai Banda on 7/13/22.
//  Copyright © 2022 Mwai Banda. All rights reserved.
//

import Foundation

final class PaymentSummaryContentViewModel: ObservableObject {
    
    @Published var selectedLabels = [ToggleLabel]()
    @Published var previousOption: ToggleOption = ToggleOption(amount: 0)
    
    /* isSelected State */
    @Published var offeringIsSelected: Bool = false
    @Published var titheIsSelected: Bool = false
    @Published var missionsIsSelected: Bool = false
    @Published var speakersIsSelected: Bool = false
    @Published var otherIsSelected: Bool = false
    
    /* Amount State */
    @Published var offeringAmount: String = "0"
    @Published var titheAmount: String = "0"
    @Published var missionsAmount: String = "0"
    @Published var speakersAmount: String = "0"
    @Published var otherAmount: String = "0"
    @Published var totalAmount: String = "0"
    
    struct ToggleOption {
        var amount: Int
        var type: ToggleLabel = .offering
    }

    
    func processToggle(isActive: Bool, type: ToggleLabel){
        if selectedLabels.contains(where: { $0 == type }) {
            let index = selectedLabels.firstIndex(where: { $0 == type}) ?? 0
            if isActive == false {
                selectedLabels.remove(at: index)
                resetPrevious()
                if selectedLabels.isEmpty {
                    resetAmounts()
                }
            }
        }else {
            selectedLabels.append(type)
            if selectedLabels.count > 1 {
                previousOption = ToggleOption(amount:  0, type: type)
                resetPrevious()
            }
            Log.d(tag: "ADD/\(type)", message: selectedLabels)
            Log.d(tag: "Previous/\(previousOption.type)", message: previousOption)
            
        }
        if selectedLabels.count == 1 {
            resetAmounts()
            switch selectedLabels.first {
            case .offering:
                offeringAmount = totalAmount
            case .tithe:
                titheAmount = totalAmount
            case .missions:
                missionsAmount = totalAmount
            case .specialSpeaker:
                speakersAmount = totalAmount
            case .other:
                otherAmount = totalAmount
                
            case .none:
                break
            }
        }
    }
    
    func processAmount(amount: String, type: ToggleLabel) {
        previousOption = ToggleOption(amount: Int(amount) ?? 0, type: type)
        switch selectedLabels.count {
        case 2:
            switch selectedLabels[0] {
                
                
            case .offering:
                switch selectedLabels[1] {
                case .offering:
                    break
                case .tithe:
                    if previousOption.type == .tithe {
                        setOfferingAmount(amount: amount)
                    } else {
                        setTitheAmount(amount: amount)
                    }
                case .missions:
                    if previousOption.type == .missions {
                        setOfferingAmount(amount: amount)
                    } else {
                        setMissionsAmount(amount: amount)
                    }
                case .specialSpeaker:
                    if previousOption.type == .specialSpeaker {
                        setOfferingAmount(amount: amount)
                    } else {
                        setSpeakersAmount(amount: amount)
                    }
                case .other:
                    if previousOption.type == .other {
                        setOfferingAmount(amount: amount)
                    } else {
                        setOtherAmount(amount: amount)
                    }
                }
                
                
            case .tithe:
                switch selectedLabels[1] {
                case .offering:
                    if previousOption.type == .offering {
                        setTitheAmount(amount: amount)
                    } else {
                        setOfferingAmount(amount: amount)
                    }
                case .tithe:
                    break
                case .missions:
                    if previousOption.type == .missions {
                        setTitheAmount(amount: amount)
                    } else {
                        setMissionsAmount(amount: amount)
                    }
                case .specialSpeaker:
                    if previousOption.type == .specialSpeaker {
                        setTitheAmount(amount: amount)
                    } else {
                        setSpeakersAmount(amount: amount)
                    }
                case .other:
                    if previousOption.type == .other {
                        setTitheAmount(amount: amount)
                    } else {
                        setOtherAmount(amount: amount)
                    }
                }
                
                
            case .missions:
                switch selectedLabels[1] {
                case .offering:
                    if previousOption.type == .offering {
                        setMissionsAmount(amount: amount)
                    } else {
                        setOfferingAmount(amount: amount)
                    }
                case .tithe:
                    if previousOption.type == .tithe {
                        setMissionsAmount(amount: amount)
                    } else {
                        setTitheAmount(amount: amount)
                    }
                case .missions:
                    break
                case .specialSpeaker:
                    if previousOption.type == .specialSpeaker {
                        setMissionsAmount(amount: amount)
                    } else {
                        setSpeakersAmount(amount: amount)
                    }
                case .other:
                    if previousOption.type == .other {
                        setMissionsAmount(amount: amount)
                    } else {
                        setOtherAmount(amount: amount)
                    }
                }
                
                
            case .specialSpeaker:
                switch selectedLabels[1] {
                case .offering:
                    if previousOption.type == .offering {
                        setSpeakersAmount(amount: amount)
                    } else {
                        setOfferingAmount(amount: amount)
                    }
                case .tithe:
                    if previousOption.type == .tithe {
                        setSpeakersAmount(amount: amount)
                    } else {
                        setTitheAmount(amount: amount)
                    }
                case .missions:
                    if previousOption.type == .missions {
                        setSpeakersAmount(amount: amount)
                    } else {
                        setMissionsAmount(amount: amount)
                    }
                case .specialSpeaker:
                    break
                case .other:
                    if previousOption.type == .other {
                        setSpeakersAmount(amount: amount)
                    } else {
                        setOtherAmount(amount: amount)
                    }
                }
                
                
            case .other:
                switch selectedLabels[1] {
                case .offering:
                    if previousOption.type == .offering {
                        setOtherAmount(amount: amount)
                    } else {
                        setOfferingAmount(amount: amount)
                    }
                case .tithe:
                    if previousOption.type == .tithe {
                        setOtherAmount(amount: amount)
                    } else {
                        setTitheAmount(amount: amount)
                    }
                case .missions:
                    if previousOption.type == .missions {
                        setOtherAmount(amount: amount)
                    } else {
                        setMissionsAmount(amount: amount)
                    }
                case .specialSpeaker:
                    if previousOption.type == .specialSpeaker {
                        setOtherAmount(amount: amount)
                    } else {
                        setSpeakersAmount(amount: amount)
                    }
                case .other:
                    break
                }
            }
            
        default:
            break
        }
    }
    func getTransactionDescription() -> String {
        if selectedLabels.count < 2 {
            var description = ""
            switch selectedLabels[0] {
            case .offering:
                description = "$\(offeringAmount): \(selectedLabels[0].rawValue)"
            case .tithe:
                description = "$\(titheAmount): \(selectedLabels[0].rawValue)"
            case .missions:
                description = "$\(missionsAmount): \(selectedLabels[0].rawValue)"
            case .specialSpeaker:
                description = "$\(speakersAmount): \(selectedLabels[0].rawValue)"
            case .other:
                description = "$\(otherAmount): \(selectedLabels[0].rawValue)"
            }
            return description
        } else {
            var description = ""
            selectedLabels.forEach { toggle in
                if selectedLabels.last == toggle {
                    switch toggle {
                    case .offering:
                        description += "$\(offeringAmount): \(toggle.rawValue)"
                    case .tithe:
                        description += "$\(titheAmount): \(toggle.rawValue)"
                    case .missions:
                        description += "$\(missionsAmount): \(toggle.rawValue)"
                    case .specialSpeaker:
                        description += "$\(speakersAmount): \(toggle.rawValue)"
                    case .other:
                        description += "$\(otherAmount): \(toggle.rawValue)"
                    }
                } else {
                    switch toggle {
                    case .offering:
                        description += "$\(offeringAmount): \(toggle.rawValue), "
                    case .tithe:
                        description += "$\(titheAmount): \(toggle.rawValue), "
                    case .missions:
                        description += "$\(missionsAmount): \(toggle.rawValue), "
                    case .specialSpeaker:
                        description += "$\(speakersAmount): \(toggle.rawValue), "
                    case .other:
                        description += "$\(otherAmount): \(toggle.rawValue), "
                    }
                    
                }
            }
            return description
        }
    }

    private func subtractAmounts(amounts: String...) -> String {
        let reminder = (Int(amounts[0]) ?? 0) - (Int(amounts[1]) ?? 0)
        return String(reminder)
    }
    
    private func addAmounts(amounts: String...) -> String {
        let reminder = (Int(amounts[0]) ?? 0) + (Int(amounts[1]) ?? 0)
        return String(reminder)
    }
    
    private func setOtherAmount(amount: String) {
        if previousOption.amount > (Int(amount) ?? 0) {
            otherAmount = addAmounts(amounts: totalAmount, amount)
        } else {
            otherAmount = subtractAmounts(amounts: totalAmount, amount)
        }
    }
    
    private func setOfferingAmount(amount: String) {
        if previousOption.amount > (Int(amount) ?? 0) {
            offeringAmount = addAmounts(amounts: totalAmount, amount)
        } else {
            offeringAmount = subtractAmounts(amounts: totalAmount, amount)
        }
    }
    
    private func setTitheAmount(amount: String) {
        if previousOption.amount > (Int(amount) ?? 0) {
            titheAmount = addAmounts(amounts: totalAmount, amount)
        } else {
            titheAmount = subtractAmounts(amounts: totalAmount, amount)
        }
    }
    
    private func setMissionsAmount(amount: String) {
        if previousOption.amount > (Int(amount) ?? 0) {
            missionsAmount = addAmounts(amounts: totalAmount, amount)
        } else {
            missionsAmount = subtractAmounts(amounts: totalAmount, amount)
        }
    }
    
    private func setSpeakersAmount(amount: String) {
        if previousOption.amount > (Int(amount) ?? 0) {
            speakersAmount = addAmounts(amounts: totalAmount, amount)
        } else {
            speakersAmount = subtractAmounts(amounts: totalAmount, amount)
        }
    }
    
    
    private func resetPrevious() {
        switch previousOption.type {
        case .offering:
            offeringAmount = "0"
        case .tithe:
            titheAmount = "0"
        case .missions:
            missionsAmount = "0"
        case .specialSpeaker:
            speakersAmount = "0"
        case .other:
            otherAmount = "0"
        }
    }
    
    private func resetAmounts() {
        offeringAmount = "0"
        titheAmount = "0"
        missionsAmount = "0"
        speakersAmount = "0"
        otherAmount = "0"
    }
}

File: MomentumAndroid/src/main/java/com/mwaibanda/momentum/android/presentation/payment/PaymentSummaryContentViewModel.kt
package com.mwaibanda.momentum.android.presentation.payment

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.mwaibanda.momentum.android.presentation.payment.PaymentSummaryContentViewModel.ToggleLabel.MISSIONS
import com.mwaibanda.momentum.android.presentation.payment.PaymentSummaryContentViewModel.ToggleLabel.OFFERING
import com.mwaibanda.momentum.android.presentation.payment.PaymentSummaryContentViewModel.ToggleLabel.OTHER
import com.mwaibanda.momentum.android.presentation.payment.PaymentSummaryContentViewModel.ToggleLabel.SPECIAL_SPEAKER
import com.mwaibanda.momentum.android.presentation.payment.PaymentSummaryContentViewModel.ToggleLabel.TITHE

class PaymentSummaryContentViewModel: ViewModel() {
    enum class ToggleLabel {
        OFFERING, TITHE, MISSIONS, SPECIAL_SPEAKER, OTHER
    }
    data class ToggleOption(val amount: Int, val type: ToggleLabel)
    var selectedLabels: MutableList<ToggleLabel> = mutableStateListOf()
    var previousOption: ToggleOption? by mutableStateOf(null)
    /* isSelected State */
    var offeringIsSelected by mutableStateOf(false)
    var titheIsSelected by mutableStateOf(false)
    var missionsIsSelected by mutableStateOf(false)
    var speakersIsSelected by mutableStateOf(false)
    var otherIsSelected by mutableStateOf(false)

    /* Amount State */
    var offeringAmount by mutableStateOf("0")
    var titheAmount by mutableStateOf("0")
    var missionsAmount by mutableStateOf("0")
    var speakersAmount by mutableStateOf("0")
    var otherAmount by mutableStateOf("0")
    var totalAmount by mutableStateOf("0")

    fun processToggle(isActive: Boolean, type: ToggleLabel){
        if (selectedLabels.contains(type)) {
            val index = selectedLabels.indexOf(type)
            if (!isActive) {
                selectedLabels.removeAt(index)
                resetPrevious()
                if (selectedLabels.isEmpty())
                    resetAmounts()
            }
        } else {
            selectedLabels.add(type)
            if (selectedLabels.count() == 1) {
                previousOption = ToggleOption(amount = 0, type = type)
                resetPrevious()
            }
        }
        if (selectedLabels.count() == 1) {
            resetAmounts()
            when (selectedLabels.first()) {
                OFFERING -> offeringAmount = totalAmount
                TITHE -> titheAmount = totalAmount
                MISSIONS -> missionsAmount = totalAmount
                SPECIAL_SPEAKER -> speakersAmount = totalAmount
                OTHER -> otherAmount = totalAmount
            }
        }
    }
    fun processAmount(amount: String, type: ToggleLabel){
        previousOption = ToggleOption(amount = amount.toInt(), type = type)
        if (selectedLabels.count() == 2) {
            when (selectedLabels[0]) {
                OFFERING -> {
                    when (selectedLabels[1]) {
                        OFFERING -> {}
                        TITHE -> {
                            if (previousOption?.type == TITHE)
                                setOffering(amount = amount)
                            else
                                setTithe(amount = amount)
                        }
                        MISSIONS -> {
                            if (previousOption?.type == MISSIONS)
                                setOffering(amount = amount)
                            else
                                setMissions(amount = amount)
                        }
                        SPECIAL_SPEAKER -> {
                            if (previousOption?.type == SPECIAL_SPEAKER)
                                setOffering(amount = amount)
                            else
                                setSpeakers(amount = amount)
                        }
                        OTHER -> {
                            if (previousOption?.type == OTHER)
                                setOffering(amount = amount)
                            else
                                setOther(amount = amount)
                        }
                    }
                }
                TITHE -> {
                    when (selectedLabels[1]) {
                        OFFERING -> {
                            if (previousOption?.type == OFFERING)
                                setTithe(amount = amount)
                            else
                                setOffering(amount = amount)
                        }
                        TITHE -> {}
                        MISSIONS -> {
                            if (previousOption?.type == MISSIONS)
                                setTithe(amount = amount)
                            else
                                setMissions(amount = amount)
                        }
                        SPECIAL_SPEAKER -> {
                            if (previousOption?.type == SPECIAL_SPEAKER)
                                setTithe(amount = amount)
                            else
                                setSpeakers(amount = amount)
                        }
                        OTHER -> {
                            if (previousOption?.type == OTHER)
                                setTithe(amount = amount)
                            else
                                setOther(amount = amount)
                        }
                    }
                }
                MISSIONS -> {
                    when (selectedLabels[1]) {
                        OFFERING ->  {
                            if (previousOption?.type == OFFERING)
                                setMissions(amount = amount)
                            else
                                setOffering(amount = amount)
                        }
                        TITHE -> {
                            if (previousOption?.type == TITHE)
                                setMissions(amount = amount)
                            else
                                setTithe(amount = amount)
                        }
                        MISSIONS -> {}
                        SPECIAL_SPEAKER -> {
                            if (previousOption?.type == SPECIAL_SPEAKER)
                                setMissions(amount = amount)
                            else
                                setSpeakers(amount = amount)
                        }
                        OTHER -> {
                            if (previousOption?.type == OTHER)
                                setMissions(amount = amount)
                            else
                                setOther(amount = amount)
                        }
                    }
                }
                SPECIAL_SPEAKER -> {
                    when (selectedLabels[1]) {
                        OFFERING -> {
                            if (previousOption?.type == OFFERING)
                                setSpeakers(amount = amount)
                            else
                                setOffering(amount = amount)
                        }
                        TITHE -> {
                            if (previousOption?.type == TITHE)
                                setSpeakers(amount = amount)
                            else
                                setTithe(amount = amount)
                        }
                        MISSIONS -> {
                            if (previousOption?.type == MISSIONS)
                                setSpeakers(amount = amount)
                            else
                                setMissions(amount = amount)
                        }
                        SPECIAL_SPEAKER -> {}
                        OTHER -> {
                            if (previousOption?.type == OTHER)
                                setSpeakers(amount = amount)
                            else
                                setOther(amount = amount)
                        }
                    }
                }
                OTHER -> {
                    when (selectedLabels[1]) {
                        OFFERING -> {
                            if (previousOption?.type == OFFERING)
                                setOther(amount = amount)
                            else
                                setOffering(amount = amount)
                        }
                        TITHE -> {
                            if (previousOption?.type == TITHE)
                                setOther(amount = amount)
                            else
                                setTithe(amount = amount)
                        }
                        MISSIONS -> {
                            if (previousOption?.type == MISSIONS)
                                setOther(amount = amount)
                            else
                                setMissions(amount = amount)
                        }
                        SPECIAL_SPEAKER -> {
                            if (previousOption?.type == SPECIAL_SPEAKER)
                                setOther(amount = amount)
                            else
                                setSpeakers(amount = amount)
                        }
                        OTHER -> {}
                    }
                }
            }
        }
    }
    fun  getTransactionDescription(otherLabel: String): String {
        if (selectedLabels.count() == 1) {
            return when(selectedLabels[0]) {
                OFFERING -> "$$offeringAmount: Offering"
                TITHE -> "$$titheAmount: Tithe"
                MISSIONS -> "$$missionsAmount: Missions"
                SPECIAL_SPEAKER -> "$$speakersAmount: Special Speaker"
                OTHER -> "$$otherAmount: ${otherLabel.ifEmpty { "Other" }}"
            }
        } else {
            var description = ""
            selectedLabels.forEach { toggleLabel ->
                if (selectedLabels.last() == toggleLabel) {
                      description += when(toggleLabel) {
                          OFFERING -> "$$offeringAmount: Offering"
                          TITHE -> "$$titheAmount: Tithe"
                          MISSIONS -> "$$missionsAmount: Missions"
                          SPECIAL_SPEAKER -> "$$speakersAmount: Special Speaker"
                          OTHER -> "$$otherAmount: ${otherLabel.ifEmpty { "Other" }}"
                      }
                } else {
                    description += when(toggleLabel) {
                        OFFERING -> "$$offeringAmount: Offering, "
                        TITHE -> "$$titheAmount: Tithe, "
                        MISSIONS -> "$$missionsAmount: Missions, "
                        SPECIAL_SPEAKER -> "$$speakersAmount: Special Speaker, "
                        OTHER -> "$$otherAmount: ${otherLabel.ifEmpty { "Other" }}"
                    }
                }
            }
            return description
        }
    }
    private fun subtractAmounts(vararg amounts: String): String {
        val reminder = amounts[0].toInt() - amounts[1].toInt()
        return reminder.toString()
    }

    private fun addAmounts(vararg amounts: String): String {
        val reminder = amounts[0].toInt() + amounts[1].toInt()
        return reminder.toString()
    }

    private fun setOther(amount: String) {
        if ((previousOption?.amount ?: 0) > amount.toInt()) {
            otherAmount = addAmounts(totalAmount, amount)
        } else {
            otherAmount = subtractAmounts(totalAmount, amount)
        }
    }

    private fun setOffering(amount: String) {
        if ((previousOption?.amount ?: 0) > amount.toInt()) {
            offeringAmount = addAmounts(totalAmount, amount)
        } else {
            offeringAmount = subtractAmounts(totalAmount, amount)
        }
    }

    private fun setTithe(amount: String) {
        if ((previousOption?.amount ?: 0) > amount.toInt()) {
            titheAmount = addAmounts(totalAmount, amount)
        } else {
            titheAmount = subtractAmounts(totalAmount, amount)
        }
    }

    private fun setMissions(amount: String) {
        if ((previousOption?.amount ?: 0) > amount.toInt()) {
            missionsAmount = addAmounts(totalAmount, amount)
        } else {
            missionsAmount = subtractAmounts(totalAmount, amount)
        }
    }

    private fun setSpeakers(amount: String) {
        if ((previousOption?.amount ?: 0) > amount.toInt()) {
            speakersAmount = addAmounts(totalAmount, amount)
        } else {
            speakersAmount = subtractAmounts(totalAmount, amount)
        }
    }

    private fun  resetPrevious(){
        when(previousOption?.type) {
            OFFERING -> offeringAmount = "0"
            TITHE -> titheAmount = "0"
            MISSIONS -> missionsAmount = "0"
            SPECIAL_SPEAKER -> speakersAmount = "0"
            OTHER -> otherAmount = "0"
            null -> {}
        }
    }

    private fun resetAmounts() {
        offeringAmount = "0"
        titheAmount = "0"
        missionsAmount = "0"
        speakersAmount = "0"
        otherAmount = "0"
    }
}